examples_x035/usbpd_sink/usbpd.h

```diff
@@ -0,0 +1,921 @@
+ * @param  None
+ * @return USBPD_BUSY if negotiation is in progress, eUSBPD_OK if successful, or an error code
+ */
+USBPD_Result_e USBPD_SinkNegotiate( void );
+
+/**
+ * @brief  Reset the USB PD module
+ * @param  None
+ * @return None
+ */
+void USBPD_Reset( void );
+
+/**
+ * @brief  Get the current state of the USB PD module
+ * @param  None
+ * @return USBPD_State_e representing the current state of the module
+ */
+USBPD_State_e USBPD_GetState( void );
+
+/**
+ * @brief  Convert USB PD state to string
+ * @param state: USBPD_State_e to convert
+ * @return Pointer to a string representing the state
+ */
+const char *USBPD_StateToStr( USBPD_State_e state );
+
+/**
+ * @brief  Convert USB PD result to string
+ * @param  result: USBPD_Result_e to convert
+ * @return Pointer to a string representing the result
+ */
+const char *USBPD_ResultToStr( USBPD_Result_e result );
+
+/**
+ * @brief  Select a new Power Data Object (PDO) to be used. No re-negotiation is needed.
+ * @param  index: Index of the PDO to select (0-based)
+ * @param  voltageIn100mV: Desired output voltage in 100mV units (e.g., 500 for 5V) (only applicable for PPS)
+ * @return USBPD_Result_e
+ */
+USBPD_Result_e USBPD_SelectPDO( uint8_t index, uint32_t voltageIn100mV );
+
+/**
+ * @brief  Get the capabilities of the USB PD Source
+ * @param[out] capabilities: Pointer to a pointer where the capabilities message structure is stored
+ * @return Number of Power Data Objects (PDOs) in the capabilities message
+ */
+size_t USBPD_GetCapabilities( USBPD_SPR_CapabilitiesMessage_t **capabilities );
+
+/**
+ * @brief  Check if the Power Data Object is a Programmable Power Supply (PPS)
+ * @param[in] pdo: Pointer to the Power Data Object to check
+ * @return true if the PDO is a PPS, false otherwise
+ */
+bool USBPD_IsPPS( const USBPD_SourcePDO_t *pdo );
+
+/**
+ * @brief  Get the USB PD Specification Revision
+ * @param  None
+ * @return USBPD_SpecificationRevision_e representing the USB PD specification revision
+ */
+USBPD_SpecificationRevision_e USBPD_GetVersion( void );
+
+#if defined( USBPD_IMPLEMENTATION )
+
+#include <string.h>
+
+typedef struct
+{
+	uint32_t ccCount;
+	volatile USBPD_State_e state;
+	USBPD_SpecificationRevision_e pdVersion;
+	USBPD_CC_e lastCCLine;
+	USBPD_SPR_CapabilitiesMessage_t caps;
+	uint8_t messageID;
+	uint8_t pdoCount;
+	bool gotSourceGoodCRC;
+} USBPD_Instance_t;
+
+static __attribute__( ( aligned( 4 ) ) ) uint8_t s_buffer[34];
+static USBPD_Instance_t s_instance = {
+	.pdVersion = eUSBPD_REV_30,
+};
+
+static USBPD_CC_e GetActiveCCLine( void );
+static void SwitchRXMode( void );
+static void SendMessage( uint8_t size );
+static void ParsePacket( void );
+
+USBPD_Result_e USBPD_Init( USBPD_VCC_e vcc )
+{
+	RCC->APB2PCENR |= RCC_APB2Periph_GPIOC | RCC_APB2Periph_AFIO;
+	RCC->AHBPCENR |= RCC_USBPD;
+
+	GPIOC->CFGHR &= ~( 0xf << ( ( 14 & 7 ) << 2 ) );
+	GPIOC->CFGHR |= ( GPIO_Speed_10MHz | GPIO_CNF_OUT_OD ) << ( ( 14 & 7 ) << 2 );
```

Comment:
What is your thinking about doing this, verses `funPinMode`? If you do use `funPinMode` it makes it a lot easier if someone wanted to port it to another system.

Reply:
Looks like the funPinMode macros are not correctly defined on x035, continuing on discord.

Reply:
Totally understood.